### PR TITLE
Create and upsert into account api tokens table

### PR DESCRIPTION
**Pull request by niki-joanne**
## Changes made in this PR
- Created account_api_tokens table in db
- Implemented Create/Update Account API token in store layer
- Created AccountAPIToken structure in domain layer

## Corresponding noted Ruby branch
https://github.com/ParamountDataManagement/paramount-api/tree/niki/golang-account-api-impl

## TODO - changes that needs to be made
Following review completion, implement DeleteToken in store layer
                                                                                        
## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [x] Defined gRPC API in proto file
- [x] regenerated go code using `make proto_generate` command

#### API implementation
- [x] Implemented API in corresponding gRPC server
- [x] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [x] Added new query in dbstore layer
- [x] Added tests in db store layer
                                                                                       
- [x] Implemented functionality in Domain Service Layer
- [x] Added test in Domain Service Layer

- [x] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

### CODE REVIEW

File: internal/domain/utils/tokenGenerator.go [hunk 1]
```diff
@@ -47,3 +48,21 @@ func GenerateRandomStringURLSafe(n int) (string, error) {
 	b, err := GenerateRandomBytes(n)
 	return base64.URLEncoding.EncodeToString(b), err
 }
+
+// GenerateRandomString returns a securely generated random string.
+// It will return an error if the system's secure random
+// number generator fails to function correctly, in which
+// case the caller should not continue.
+func GenerateRandomString(n int) (string, error) {
```
Comment (lhan-the-real) [inline internal/domain/utils/tokenGenerator.go:56]: What's the Ruby's definition of following functionality?
```
SecureRandom.alphanumeric(64)
```

Let's research on what does the above function mean in Ruby, and based on that definition, define our helper function's expected output.

And add tests on this function.

Comment (niki-joanne) [reply internal/domain/utils/tokenGenerator.go]: I did find a secure-random library in golang, but it didn't include an implementation for a random _alphanumeric_ string, so I searched for alternatives, which is what led me to this implementation

Found another library, go-random, which has a function to support secure random alphanumeric string generation, and I've used that now. Please verify.

File: internal/domain/utils/utils.go [hunk 2]
```diff
@@ -46,6 +46,14 @@ func GenerateVerificationToken() (string, error) {
 	return token, nil
 }
 
+func GenerateAccountAPIToken() (string, error) {
+	token, err := GenerateRandomAlphanumericString(64)
+	if err != nil {
+		return "", pdmerr.NewParent(err, "failed to generate account API token")
+	}
+	return token, err
```
Comment (lhan-the-real) [inline internal/domain/utils/utils.go:54]: return token, nil

File: internal/domain/utils/tokenGenerator.go [hunk 3]
```diff
@@ -47,3 +50,19 @@ func GenerateRandomStringURLSafe(n int) (string, error) {
 	b, err := GenerateRandomBytes(n)
 	return base64.URLEncoding.EncodeToString(b), err
 }
+
+// GenerateRandomString returns a securely generated random string.
+// It will return an error if the system's secure random
+// number generator fails to function correctly, in which
+// case the caller should not continue.
+func GenerateRandomAlphanumericString(n int) (string, error) {
```
Comment (lhan-the-real) [inline internal/domain/utils/tokenGenerator.go:58]: Love this finding!

File: internal/domain/accounts/accountsStore.go [hunk 4]
```diff
@@ -166,8 +166,34 @@ func (a *accountsStore) ResetAccount(ctx context.Context, accountID int32) (Acco
...
+		on conflict (account_id) do update
+			set token = excluded.token
+	returning id, updated_at, account_id, token;`)
+
+	err = a.db.QueryRowxContext(
+		ctx,
+		query,
+		accountID,
+		tokenValue).StructScan(&createdAccountAPIToken)
+
```
Comment (lhan-the-real) [inline internal/domain/accounts/accountsStore.go:191]: could you explain this query a bit? I'm confused about where is the excluded.token come from

File: internal/domain/accounts/accountsStore.go [hunk 5]
```diff
@@ -23,7 +23,7 @@ type AccountsStore interface {
 	DeleteAccount(ctx context.Context, accountID int32) error
 	UpdateAccount(ctx context.Context, updateInfoFields []string, updateAccountInfo Account) (updatedAccount Account, err error)
 	ResetAccount(ctx context.Context, accountID int32) (account Account, err error)
-	CreateAPIToken(ctx context.Context) error
+	CreateAPIToken(ctx context.Context, accountID int32) (createdAccountAPIToken *AccountAPIToken, err error)
```
Comment (lhan-the-real) [inline internal/domain/accounts/accountsStore.go:26]: Let's rename it to `UpsertAPIToken` to reflect the behavior in the implementation.

File: internal/domain/accounts/accountsStore.go (hunk 4 above)
Comment (niki-joanne) [reply internal/domain/accounts/accountsStore.go:191]: `excluded` references the "excluded" row that wasn't inserted in the table, as a result of the conflict
So, in the event of a conflict, we set the token value in that row to that of the attempted entry, which we reference through `excluded`

Comment (lhan-the-real) [reply internal/domain/accounts/accountsStore.go:191]: Nice! Love it! Thanks for sharing!

**Review by lhan-the-real — APPROVED** [review]

**Pull request merged**